src/ui/shortcutssettings.cc

```diff
@@ -0,0 +1,230 @@
+      });
+      context_menu_->addSeparator();
+      for (auto& shortcut : item_shortcuts) {
+        auto remove_action = new QAction(tr("Remove %1").arg(shortcut.toString()), context_menu_);
+        connect(remove_action, &QAction::triggered, [this, item_type, shortcut] () {
+          model_->removeShortcut(item_type, shortcut);
+          modified_types_.insert(item_type);
+        });
+        context_menu_->addAction(remove_action);
+      }
+      context_menu_->exec(ui->shortcutsTree->mapToGlobal(point));
+    }
+  }
+}
+
+void ShortcutsDialog::showEvent(QShowEvent* event) {
+  ui->shortcutsTree->setColumnWidth(0, ui->shortcutsTree->width()*2/3);
+  ui->filter->clear();
+  QDialog::showEvent(event);
+}
+
+void ShortcutsDialog::showShortcutSelectionDialog(ShortcutType type) {
+  shortcut_selection_->reset();
+  connect(shortcut_selection_, &ShortcutEditDialog::accepted, [this, type] () {
+    QKeySequence sequence = shortcut_selection_->getSequence();
+    if (!sequence.isEmpty()) {
+      model_->addShortcut(type, sequence);
+      modified_types_.insert(type);
+    }
+  });
+  shortcut_selection_->exec();
+}
+
+void ShortcutsDialog::showShortcutSelectionDialog(QModelIndex index) {
+  if (!index.isValid()) {
+    return;
+  }
+  auto is_category = index.data(ShortcutsModel::CATEGORY_ROLE).toBool();
+  if (is_category) {
+    return;
+  }
+  auto type = static_cast<ShortcutType>(index.data(ShortcutsModel::TYPE_ROLE).value<uint32_t>());
+  showShortcutSelectionDialog(type);
+}
+
+ShortcutEditDialog::ShortcutEditDialog(QWidget* parent) :
+    QDialog(parent, Qt::WindowTitleHint | Qt::WindowSystemMenuHint | Qt::WindowCloseButtonHint),
+    ui(new Ui::ShortcutEditDialog) {
+  ui->setupUi(this);
+  ui->conflictLabel->setStyleSheet("QLabel { color : red; }");
+  ui->conflictLabel->setWordWrap(true);
+  connect(ui->lineEdit, &ShortcutEdit::returnPressed, this, &ShortcutEditDialog::accept);
+  connect(ui->lineEdit, &ShortcutEdit::escapePressed, this, &ShortcutEditDialog::reject);
+  connect(ui->lineEdit, &ShortcutEdit::textChanged, this, &ShortcutEditDialog::checkConflicts);
+  connect(ui->altBox, &QCheckBox::stateChanged, this, &ShortcutEditDialog::checkConflicts);
+  connect(ui->ctrlBox, &QCheckBox::stateChanged, this, &ShortcutEditDialog::checkConflicts);
+  connect(ui->shiftBox, &QCheckBox::stateChanged, this, &ShortcutEditDialog::checkConflicts);
+  connect(ui->metaBox, &QCheckBox::stateChanged, this, &ShortcutEditDialog::checkConflicts);
+#ifndef Q_OS_MAC
+  ui->metaBox->setVisible(false);
+#endif
+}
+
+ShortcutEditDialog::~ShortcutEditDialog() {
+  delete ui;
+}
+
+void ShortcutEditDialog::reset() {
+  ui->altBox->setChecked(false);
+  ui->ctrlBox->setChecked(false);
+  ui->shiftBox->setChecked(false);
+  ui->metaBox->setChecked(false);
+  ui->lineEdit->reset();
+  ui->conflictLabel->clear();
+  disconnect();
+}
+
+QKeySequence ShortcutEditDialog::getSequence() {
+  int key = ui->lineEdit->key();
```

Comment:
I would agree if not for the fact that Qt itself uses ints for it - so if it will bite us we won't be able to do anything about it either way (`key()` returns int) :P